src/Interfaces/FwupdManager.vala

```diff
@@ -0,0 +1,317 @@
+/*
+* Copyright (c) 2020 elementary, Inc. (https://elementary.io)
+*
+* This program is free software; you can redistribute it and/or
+* modify it under the terms of the GNU General Public
+* License as published by the Free Software Foundation; either
+* version 3 of the License, or (at your option) any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+* General Public License for more details.
+*
+* You should have received a copy of the GNU General Public
+* License along with this program; if not, write to the
+* Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+* Boston, MA 02110-1301 USA
+*
+* Authored by: Marius Meisenzahl <mariusmeisenzahl@gmail.com>
+*/
+
+extern int ro_fd (string path);
+
+public class About.FwupdManager : Object {
+    [DBus (name = "org.freedesktop.fwupd")]
+    public interface FwupdInterface : Object {
+        public abstract signal void device_added (GLib.HashTable<string, Variant> device);
+        public abstract signal void device_removed (GLib.HashTable<string, Variant> device);
+
+        public abstract async GLib.HashTable<string, Variant>[] get_devices () throws GLib.Error;
+        public abstract async GLib.HashTable<string, Variant>[] get_releases (string device_id) throws GLib.Error;
+        public abstract async void install (string id, UnixInputStream handle, GLib.HashTable<string, Variant> options) throws GLib.Error;
+        public abstract async GLib.HashTable<string, Variant>[] get_details (UnixInputStream handle) throws GLib.Error;
+    }
+
+    private FwupdInterface fwupd;
+
+    public signal void on_device_added (Device device);
+    public signal void on_device_error (Device device, string error);
+    public signal void on_device_removed (Device device);
+
+    static FwupdManager? instance = null;
```

Comment:
I think I'd prefer to not use a singleton for this class. It would be better to construct one instance of this class in `FirmwareView` and then use signals from `FirmwareUpdateRow` to trigger the update process back in `FirmwareView`. Or, if that turns out to be too complex, you could pass the reference from `FirmwareView` to `FirmwareUpdateRow`

Reply:
Or, if you want to efficiently check it's the same file and not download it again, you could check the timestamps like what AppCenter does in the screenshot cache.